### DISCUSSION TITLE
How do I manage a proxy server from within another Python app?

### DISCUSSION THREAD

**Discussion by JJTech0130** [mitmproxy/mitmproxy#6903] [Q&A]
I would really like to be able to start, configure, and manage an HTTP proxy from within my Python application (built with Quart).

I don't think I can get what I want done with the `DumpManager` API, as it was giving me trouble and not giving me control over its execution (for example, it had trouble when the event loop changed out from under it due to Quart hot-reloading).

Instead, I'm trying to directly use the `mitmproxy.proxy` API, as exemplified by the test in `mitmproxy.proxy.server`. However, that only shows a "reverse" proxy, and I'm looking for a "normal" HTTP proxy and can't find any documentation on that.

Are there any existing projects that implement this, or other places I could look for help? If not, what exactly do I need to setup to make this work: I assume I need to change the layers in `layer_stack`, but I'm not sure what layers I would need.

**Comment by emanuele-em** [comment]
I thing [Addons](https://docs.mitmproxy.org/stable/addons-overview/) are the solution, with addons you can include mitmproxy as a normal python library

**Comment by JJTech0130** [comment]
The problem with Addons is that means that my app is driven by mitmproxy, right? As in, I would have to start `mitmdump` or another tool and it would happen to load my app.
Instead, I want my app to be the primary thing (it can do a lot of its function w/o `mitmproxy` and be able to start mitmproxy inside it when it wants to.

**Reply by Prinzhorn** [reply]
You can spawn a `mitmdump` process from your app and use some form of IPC to communicate with an addon.

**Reply by JJTech0130** [reply]
Sure, but it would be a lot easier if I could just call whatever underlying API directly, considering my app is already in Python

**Reply by Prinzhorn** [reply]
Sure, but that's not a use-case we currently support.
